k2/csrc/fsa_algo.cc

```diff
@@ -0,0 +1,149 @@
+// k2/csrc/fsa_algo.cc
+
+// Copyright (c)  2020  Fangjun Kuang (csukuangfj@gmail.com)
+
+// See ../../LICENSE for clarification regarding multiple authors
+
+#include "k2/csrc/fsa_algo.h"
+
+#include <stack>
+
+#include "glog/logging.h"
+
+namespace {
+
+// depth first search state
+struct DfsState {
+  int32_t s;      // state number of the visiting node
+  int32_t begin;  // arc index of the visiting arc
+  int32_t end;    // end of the arc index of the visiting node
+};
+
+}  // namespace
+
+namespace k2 {
+
+// The implementation of this function is inspired by
+// http://www.openfst.org/doxygen/fst/html/connect_8h_source.html
+void ConnectCore(const Fsa &fsa, std::vector<int32_t> *state_map) {
+  CHECK_NOTNULL(state_map);
+
+  state_map->clear();
+  if (IsEmpty(fsa)) return;
+
+  auto num_states = fsa.NumStates();
+  auto final_state = num_states - 1;
+
+  std::vector<bool> accessible(num_states, false);
+  std::vector<bool> coaccessible(num_states, false);
+  std::vector<bool> visited(num_states, false);
+
+  accessible.front() = true;
+  coaccessible.back() = true;
+
+  std::stack<DfsState> stack;
+  stack.push({0, fsa.arc_indexes[0], fsa.arc_indexes[1]});
+  visited[0] = true;
+
+  while (!stack.empty()) {
+    auto &top = stack.top();
+
+    if (top.begin == top.end) {
```

Comment:
`curr_state.arc_begin == curr_state.arc_end` may be more readable, just a suggestion, do what you prefer.

Reply:
Done.